### DISCUSSION TITLE
Why when I run yarn build it always build even if nothing has changed?

### DISCUSSION THREAD

**Discussion by Hideman85** [ojkelly/yarn.build#169] [Q&A] [answered]
Hi everyone,
I seen that yarn build should keep track of what has been built but if I run two consecutive yarn build it rebuild all.
Could I get to know where does it store the tracking information of what has been built?
Thanks in advance for your help :)

**Comment by ojkelly** [comment, marked as answer]
Great question!

The state of what's build is tracked here `.yarn/yarn.build.json`. Inside there, we track each command thats run. So `yarn build` and `yarn test` are tracked separately.

If any command re-runs if nothings changed, it means something did change. We look at the whole folder and check the last modified date for each file.

Typically this means output files are not being tracked and are being picked up as input. We try to guess based on node/npm conventions where your input and output files are.

You can explicitly define the input or output folders in `package.json` with:  

```
{
  "name": "@internal/lorem-ipsum",
  "version": "1.0.0",
  "license": "UNLICENSED",
  "private": true,
  "scripts": {
    "build": "tsc -outDir dist",
    "test": "jest"
  },
  "yarn.build": {
    "input": "src",
    "output": "dist"
  }
}
```

The definition for the config is here https://github.com/ojkelly/yarn.build/blob/main/packages/plugins/shared/src/config.ts#L11

Input and output can also be folders, if you're setup is a bit more complex:

```
  "yarn.build": {
    "input": [
      "src",
      "tests"
    ],
    "output": [
      "dist",
      "coverage"
    ]
  }
````

**Reply by Hideman85** [reply]
Super thank you so much for your response.
That was the exact issue when I added the explicit input/output it did work.

Then I can advise the following improvements:
- It would be great if the command is just `tsc` to parse the `.tsconfig.json` to get to know in/out
- The next one is: I build once, then I deleted the out folder and it did not build again. Then a test if the out folder is existing would be nice too

Please let me know if that make sense or if I need to configure something else?

**Reply by ojkelly** [reply]
Parsing `tsconfig.json` for in/out is a great idea. I'll get a ticket added for that.

A positive check for the existence of the output folder, when it's explicitly known, is also a great idea. We can't do it when there's no explicit indication of what the output folder is, but it makes sense to add for those situations where we know it.

**Reply by ojkelly** [reply]
#171 and #170

**Reply by Hideman85** [reply]
Great, looking forward to it :slightly_smiling_face:

**Comment by jeffrson** [comment]
> We try to guess based on node/npm conventions where your input and output files are.

Hmmm, I wonder, how it is done.
My input files are below "src" (isn't it some "convention"?) - yarn.build always rebuilds until I set this inside "yarn.build" in package.json as input:
```
"yarn.build": {
    "input": "src",
    "output": "dist"
  }
```
"output" doesn't matter, as it seems.

From a quick look at the source it seems, default for "input" is ".". Isn't this a bad choice given building always should change something below "."?
